src/cli/self_update.rs

```diff
@@ -390,13 +397,12 @@ fn check_existence_of_rustc_or_cargo_in_path(no_prompt: bool) -> Result<()> {
         err!("Otherwise you may have confusion unless you are careful with your PATH");
         err!("If you are sure that you want both rustup and your already installed Rust");
         err!("then please restart the installation and pass `-y' to bypass this check.");
-        Err("cannot install while Rust is installed".into())
-    } else {
-        Ok(())
+        ignorable_error("cannot install while Rust is installed".into(), no_prompt)?;
```

Comment:
Errata: this is *not* ignorable if the location of the existing rust is in the CARGO_BIN directory that rustup's proxies will be written to. We're making this easier to run into, so perhaps we should check for that at this point.

Reply:
Should it be possible to skip it in any way (e.g. environment variable) or not?

Reply:
Currently the code will give a warning later about "tool X is already installed", and not touch those binaries. This would mean that rustup isn't actually installed. I don't think it makes sense to allow that to be bypassed, but would like @kinnison 's input. This might be something to do as a separate PR for instance.

Reply:
The cases this PR covers are to do with `rustc` on the `PATH` or about detecting an old-style `rustup.sh` install as I understand the change.  If it happens that it's in `CARGO_BIN` then the user is going to get warnings as you say, and if they ignore those then they get to keep all the pieces.  I don't think we can stop *all* the footguns -- this is simply a way to allow another footgun for someone who really can't install `rustup` otherwise due to immutable underlying layers.

Reply:
The concern I have is that this is escalating it from a read-the-docs situation to a y/n prompt, so perhaps a lot easier to perform targeted foot removal.

Reply:
Possibly so.  I shall check the feel of that in my testing which I'm about to undertake.

Reply:
Okay so the issue here is that if the `rustc/cargo` is in `$CARGO_BIN` then the only real pain is that `rustup` will replace them with proxies without prompting.  The only tools you get the warnings about are `rustfmt` and `cargo-fmt` because those used to be provided via `cargo install` rather than always being proxied via `rustup`.  I think that the footgun is permissible in this case because it's *exceedingly* unlikely that the user will have set up this kind of thing without also being aware of what `rustup-init` will end up doing